src/DataflowPass.cpp

```diff
@@ -168,79 +205,132 @@ void DataflowPass::transfer(Instruction *instruction,
+                dynamic_cast<ParameterAnnotation *>(mayParamAnnoWithField)) {
+          logout("found param struct annotation "
+                 << paramAnno->generateStringRep());
+          this->insertAnnotation(paramAnno, pv);
+          continue;
+        }
+      }
 
-          // if no annotations, treat it as unknown function
-          MethodsSet *methods =
-              inputProgramPoint.getPVRef(arg, false)->getMethodsSetRef();
-      logout("GETMAIN-6 ON '" << arg
-                              << "'") this->onUnknownFunctionCall(methods);
+      // no annotations found, treat function call as unknown function
+      logout("no annotations found for " << fnName);
+      MethodsSet *methods = pv->getMethodsSetRef();
+      this->onUnknownFunctionCall(methods);
     }
   } else if (AllocaInst *allocate = dyn_cast<AllocaInst>(instruction)) {
 
     // searches for struct annotations
-    if (llvm::StructType *structType =
-            llvm::dyn_cast<llvm::StructType>(allocate->getAllocatedType())) {
-      std::string structName = structType->getName();
-      structName = dataflow::sliceString(
-          structName, structName.find_last_of('.') + 1, structName.size() - 1);
-      int numFields = structType->getNumElements();
-      for (int i = 0; i < numFields; i++) {
-
-        Annotation *anno = this->annotations.getStructAnnotation(structName, i);
-        if (StructAnnotation *structAnno =
-                dynamic_cast<StructAnnotation *>(anno)) {
-          std::set<std::string> structMethods =
-              structAnno->getAnnotationMethods();
-          AnnotationType structAnnoType = structAnno->getAnnotationType();
-          ProgramVariable sourceVar = ProgramVariable(allocate, i);
-
-          std::string arg = sourceVar.getCleanedName();
-          if (!sourceVar.isIdentifier()) {
-            continue;
-          }
+    StructType *structType =
+        llvm::dyn_cast<llvm::StructType>(allocate->getAllocatedType());
 
-          logout("found annotation "
-                 << dataflow::setToString(structMethods) << " "
-                 << annotationTypeToString(structAnnoType)
-                 << " for field index " << std::to_string(i) << " var name "
-                 << sourceVar.getRawName())
-
-              for (std::string methodName : structMethods) {
-            MethodsSet *methods =
-                inputProgramPoint.getPVRef(arg, false)->getMethodsSetRef();
-            logout("var = '" << arg << "' method name = '" << methodName
-                             << "'") this->onAnnotation(methods, methodName,
-                                                        structAnnoType);
-          }
+    if (!structType && allocate->getAllocatedType()->isPointerTy()) {
+      structType = dataflow::unwrapValuePointerToStruct(allocate);
+    }
+
+    if (!structType) {
+      return;
+    }
+
+    std::string structName = structType->getName();
+    structName = dataflow::sliceString(
```

Comment:
Minor, but I think `sliceString` belongs in Utils rather than dataflow.

Reply:
Probably can/should be addressed in a separate, smaller PR

Reply:
I'm thinking of just having a `Utils` and `dataflow` namespace instead of only `dataflow`. It seems minor enough to change it in this PR, though.

Reply:
> It seems minor enough to change it in this PR, though

This is a dangerous impulse that leads to PRs that are hard to review, because they contain many unrelated changes. In general, anything that can be a small, simple PR should be, so that it can be reviewed and merged quickly and easily.